#820 Make it a bit easier to run config functions without going through galsim executable

rmjarvis opened the pull request:
@esheldon reported on issue #820 that when running config functions by hand within python, e.g. `galsim.config.BuildGSObject(config, 'gal')`, it wasn't allowing the rng to be automatic for things that have `_takes_rng = True`.

This is normally not a problem when running config files through the `galsim` executable, since prior to that call, an appropriate rng will have bee constructed.  However, I think it does make sense to allow users to call functions such as this without requiring them to manually make the rng and put it in `config['rng']`.  So that's done here.

However, I also took this opportunity to go through the missing config coverage (which was a lot) and add unit tests to cover the things that weren't being covered by anything else yet.  Along the way, I found and fixed a number of bugs in edge case scenarios and made a few other changes to make it easier to call the config functions manually.  Here is a complete list of changes:
- Setting config['rng'] is no longer required when manually running commands like `galsim.config.BuildGSObject`.
- Changed the name of galsim.config.CalculateNoiseVar to the slightly more verbose, but also more readable CalculateNoiseVariance.
- Allow PoissonNoise without any sky level, in which case only the shot noise in the signal photons contribute to the noise.  Likewise for CCDNoise.
- Let 'None' in the config file mean `None`.
- Require 'max_extra_noise' to be explicitly set for photon shooting if you want it rather than have it default to 0.01.
- Added `--except_abort` option to `galsim` executable to abort execution if a file has an error, rather than just reporting the exception and continuing on (which is still the default behavior).
- Added optional probability parameter 'p' for Random bool values.
- Allow `drawImage` to take a non-uniform WCS when default constructing the image from either `nx`,`ny` or `bounds`.  (Note: this one is not a config change -- it was just a restriction that we had in `drawImage` that isn't actually necessary.)

Also a couple other functionality-neutral changes you'll notice here:
- When people used galsim.utilities for things, there was a mix of using the full `galsim.utilities.blah` vs `from . import utilities` ... `utilities.blah`.  It seemed that the former was more common, so I switched everything to use that style for consistency.
- Josh pointed out during the hack week that `_absdet` in SBTransform` wasn't actually correct.  It's really the total flux scaling.  And there was another parameter called `_fluxScaling`, which is really the amplitude scaling.  So I renamed these `_fluxScaling` and `_ampScaling` respectively to be clearer.  And I kept `_absdet` to actually mean |det(m)|.
- I also made SBConvolve and SBTransform delay some calculations until they are actually needed.  This won't usually be a performance change, since you usually do need them eventually when you draw the object, but for my `test_reject` function that builds a lot of objects and never gets around to drawing them, it dropped the running time from 7.5 seconds to a little over 2 seconds.

rmandelb commented:
I'm getting some output from the unit tests on my Mac:
```
Traceback (most recent call last):
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/site-packages/galsim/config/process.py", line 822, in worker
    ps = pstats.Stats(pr, stream=s).sort_stats(sortby).reverse_order()
  File "/Library/Frameworks/EPD64.framework/Versions/7.2/lib/python2.7/pstats.py", line 233, in sort_stats
    sort_tuple = sort_tuple + sort_arg_defs[word][0]
KeyError: 'tottime'
Process Process-4:
```

and on the Linux cluster, something completely different:
```
ERROR: Test the options in config to whiten images
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/python27/lib/python2.7/site-packages/nose-1.3.0-py2.7.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/rmandelb/git/GalSim/tests/galsim_test_helpers.py", line 559, in f2
    result = f(*args, **kwargs)
  File "/home/rmandelb/git/GalSim/tests/test_config_noise.py", line 661, in test_whiten
    im3d, cv3d = galsim.config.BuildStamp(config2)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/stamp.py", line 405, in BuildStamp
    im, current_var = builder.addNoise(stamp,config,im,skip,current_var,logger)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/stamp.py", line 876, in addNoise
    current_var = galsim.config.AddNoise(base,image,current_var,logger)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/noise.py", line 88, in AddNoise
    var = builder.addNoise(noise, config, im, rng, current_var, draw_method, logger)
  File "/home/rmandelb/software/lib/python2.7/site-packages/galsim/config/noise.py", line 326, in addNoise
    "Whitening already added more noise than the requested Poisson noise.")
RuntimeError: Whitening already added more noise than the requested Poisson noise.
-------------------- >> begin captured stdout << ---------------------
From whiten, current_var =  9.81378306774e-05
From BuildStamp, current_var =  9.81378306774e-05
```

rmjarvis commented:
> RuntimeError: Whitening already added more noise than the requested Poisson noise.

Travis found that one too, so I was already working on it.  I'm not actually sure what the error was, but when I cleaned up that bit of code slightly to not use quite so many temporary numpy arrays, the error went away.  Weird.  cf.1b6060d

> KeyError: 'tottime'

This seems to indicate that you have an old version of pstats.  AFAICT, the name for this used to be 'time' and they switched it to 'tottime', but 'time' is still a valid alias.  So assuming this alias will remain valid going forward, I just switched to that.  cf. 7f7d909c5c

rmandelb commented:
I confirm these are both fixed, thanks.  I plan to go over the code sometime before Thanksgiving (but likely not until Weds or so).

rmandelb commented on tests/test_config_value.py:
```diff
@@ -980,6 +1100,78 @@ def test_shear_value():
+        config['gal'] = { 'redshift' : 1.3 }
+    except ImportError:
+        pass
+    nfw_halo = galsim.NFWHalo(mass=1.e14, conc=4, redshift=0.3)
```
Sorry to pick on this random example, but there are lots of magic numbers that need to be written identically between the config and the direct cases, so this is another test routine that might benefit from defining them as variables.  This is true in quite a few of the new tests.

rmjarvis replied:
OK. I did these two, but I didn't try to comprehensively go through the tests to look for examples where doing this would be helpful.  If you notice any others that you think are egregious, let me know.  (Personally, I don't much mind magic numbers in tests -- in the main code is where I try to avoid them.)

rmjarvis commented on CHANGELOG.md:
```diff
@@ -106,3 +108,17 @@ New config features
+  verbose, but also more readable CalculateNoiseVariance. (#820)
+- Setting config['rng'] is no longer required when manually running commands
+  like `galsim.config.BuildGSObject`.  If you do not care about deterministic
+  pseudo-rngs, then it will just use /dev/urandom for you. (#820)
```
Sure.  I can add that.  Although, it would require that you are running things with a logger, since that's where config warnings are emitted, and I might guess that most people who neglect to specify the rng might also neglect to pass in a logger.

rmandelb replied:
Oh.  Well, that's not terribly satisfying.

rmjarvis commented on CHANGELOG.md:
```diff
@@ -106,3 +108,17 @@ New config features
+  like `galsim.config.BuildGSObject`.  If you do not care about deterministic
+  pseudo-rngs, then it will just use /dev/urandom for you. (#820)
+- Allow PoissonNoise without any sky level, in which case only the shot noise
+  in the signal photons contribute to the noise.  Likewise for CCDNoise. (#820)
```
I agree, and that's why it wasn't there before.  But it seemed like a silly restriction to not allow using PoissonNoise without a sky.  It seems perfectly legitimate to investigate galaxies where the only noise is from their own photons.

rmjarvis commented on galsim/config/noise.py:
```diff
@@ -506,9 +490,9 @@ def addNoiseVariance(self, config, base, im, include_obj_var, logger):
             # which is the object variance in each pixel.
             if gain != 1.0:
-                im += base['current_image']/gain
+                im += base['current_noise_image']/gain
```
It's just an efficiency.  Dividing by 1 is not a trivial operation on a image.  So it's better to check first and skip the calculation if it won't actually do anything.

rmandelb replied:
OK, I wouldn't have guessed that this mattered.

rmjarvis replied:
It likely would be a very minor inefficiency, you're right.  But these little things can add up.  :)

rmjarvis commented on tests/test_config_image.py:
```diff
@@ -31,9 +33,612 @@
+def test_reject():
+    """Test various ways that objects can be rejected.
+    """
+    # Make a custom function for rejecting COSMOSCatalog objects that use Sersics with n > 2.
```
Do you think this is more helpful than the example in cgc.yaml?

rmandelb replied:
I don't think it's _more_ helpful.  I just think it's another helpful thing.  They showcase different functionality (i.e., the one in cgc.yaml is based on properties of the drawn image whereas the one in this test lets you really dig into the COSMOSCatalog, and I don't find that trivial/obvious, personally).

rmjarvis replied:
I added something [here](https://github.com/GalSim-developers/GalSim/wiki/Config-Stamp#stamp-types).  The link is to the master version, which doesn't actually have this code yet, but it will once this is merged.

rmandelb commented on galsim/config/extra.py:
```diff
@@ -164,71 +162,70 @@ def WriteExtraOutputs(config, main_data, logger=None):
-            ntries = ntries + 1
-        else:
-            ntries = 1
+    output = config['output']
```
Cool.

rmjarvis commented on examples/demo7.yaml:
```diff
@@ -191,6 +191,17 @@ stamp :
+    # but only for photon
+    max_extra_noise :
+        type: List
+        items: [ None, 0.01 ]
```
That was what I was starting to write in that sentence that I didn't finish and apparently never came back to.

rmjarvis commented:
Thanks Rachel.  I think I addressed everything.

rmjarvis commented:
Any objection to merging this?

jmeyers314 commented:
>Any objection to merging this?

I bet you were waiting to hear from me on this...  Let me give it a quick once-over.

rmjarvis commented:
:)  Yep.  Thanks.

jmeyers314 commented:
It was a quick scan, but looks good to me.  If nothing else, I learned of the existence of `pragma: no branch` and `RealGalaxyOriginal` :).

rmjarvis commented:
Great.  Thanks, Josh.
